---
title: "R Notebook"
output: html_notebook
---

# Set Global Variables here

```{r}
# Please change the working directory
Working_dir <- "~/Desktop/AVATAR/PADIT-seq-Manuscript/Code_Availability_Revision/"

# Please be sure to install these libraries
library(pheatmap); library(ggplot2); library(DESeq2); library(ggrepel); library("stringr"); library(viridis)

# FDR cutoff (do not change)
Cutoff_Value <- 0.05
```

# Important functions
```{r}
######################################################### Longest stretch of 1's
Longest_1s <- function(vec) {
  # Ensure that the vector contains only 0s and 1s
  if (!all(vec %in% c(0, 1))) {
    stop("The vector should contain only 0s and 1s")
  }

  # Use rle to compute the lengths of sequences of 1s and 0s
  runs <- rle(vec)

  # Extract the lengths of sequences of 1s only
  lengths_of_ones <- runs$lengths[runs$values == 1]

  # If there are no sequences of 1s, return a maximum length of 0 and start position as NA
  if (length(lengths_of_ones) == 0) {
    return(list(max_length = 0, start_position = NA))
  }

  # Identify the maximum length of the 1s sequences
  max_length <- max(lengths_of_ones)

  # Find the position (index) where the maximum stretch of 1s begins
  start_positions_of_ones <- cumsum(runs$lengths) - runs$lengths + 1
  start_position <- start_positions_of_ones[which(runs$lengths == max_length & runs$values == 1)[1]]

  # Return the maximum length and start position as a list
  return(list(max_length = max_length, start_position = start_position))
}

# Example usage:
vec <- c(1, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1)
result <- Longest_1s(vec)
print(paste("Max length:", result$max_length, "| Start position:", result$start_position))

```


# Read HOXD13 8-mer PADIT-seq log2foldchange values for the 3 registers
```{r}
################################################################# Read HOXD13 8-mer PADIT-seq log2foldchange values for the 3 registers
Register_1_8 <- read.table(paste(paste(Working_dir, "Output_Files", sep = "/"), "HOXD13_all8mers_1_8_median.txt", sep = "/"), header = TRUE)
Register_2_9 <- read.table(paste(paste(Working_dir, "Output_Files", sep = "/"), "HOXD13_all8mers_2_9_median.txt", sep = "/"), header = TRUE)
Register_3_10 <- read.table(paste(paste(Working_dir, "Output_Files", sep = "/"), "HOXD13_all8mers_3_10_median.txt", sep = "/"), header = TRUE)

# Re-define MPRAactivity
Register_1_8$MPRAactivity <- (Register_1_8$log2FoldChange > 0 & Register_1_8$padj < Cutoff_Value)*1
Register_2_9$MPRAactivity <- (Register_2_9$log2FoldChange > 0 & Register_2_9$padj < Cutoff_Value)*1
Register_3_10$MPRAactivity <- (Register_3_10$log2FoldChange > 0 & Register_3_10$padj < Cutoff_Value)*1

# Number of red points?
table(Register_1_8$MPRAactivity)
table(Register_2_9$MPRAactivity)
table(Register_3_10$MPRAactivity)

# Format
colnames(Register_1_8) <- paste(colnames(Register_1_8), "1_8", sep = "_")
colnames(Register_2_9) <- paste(colnames(Register_2_9), "2_9", sep = "_")
colnames(Register_3_10) <- paste(colnames(Register_3_10), "3_10", sep = "_")
TEMP1 <- merge(Register_1_8, Register_2_9,
               by.x = "X8.mer_1_8", by.y = "X8.mer_2_9")
TEMP2 <- merge(TEMP1, Register_3_10,
               by.x = "X8.mer_1_8", by.y = "X8.mer_3_10")
HOXD13_PBM_Escores_PADIT <- TEMP2; rm(TEMP1); rm(TEMP2)
HOXD13_PBM_Escores_PADIT$Mean_log2FC <- apply(HOXD13_PBM_Escores_PADIT[,c("log2FoldChange_1_8", "log2FoldChange_2_9", "log2FoldChange_3_10")], 1, median)
HOXD13_PBM_Escores_PADIT$MPRAactivity <- apply(HOXD13_PBM_Escores_PADIT[,c("MPRAactivity_1_8", "MPRAactivity_2_9", "MPRAactivity_3_10")], 1, median)

#
df1 <- HOXD13_PBM_Escores_PADIT[,c("X8.mer_1_8", "Mean_log2FC", "MPRAactivity")]
colnames(df1) <- c("TFBS", "log2FoldChange", "MPRAactivity")
df2 <- HOXD13_PBM_Escores_PADIT[,c("X8.mer.1_1_8", "Mean_log2FC", "MPRAactivity")]
colnames(df2) <- c("TFBS", "log2FoldChange", "MPRAactivity")
dds_results_HOXD13_df <- rbind(df1, df2)

```



# Tile PADIT-seq HOXD13 8-mers on HT-SELEX reads
```{r}
for(Select_cycle in c("input", "cycle1", "cycle2", "cycle3", "cycle4"))
{
  # Read HOXD13 SELEX-seq DNA sequences
  if(Select_cycle == "input")
  {
    HOXD13_CHIP_tmp <- unique(read.table(paste(paste(Working_dir, "Input_Files/HT-SELEX_data_HOXD13", sep = "/"), "ZeroCycle_TGATAA20NATA_0_0.fastq.txt", sep = "/"), header = FALSE))
  }
  if(Select_cycle == "cycle1")
  {
    HOXD13_CHIP_tmp <- unique(read.table(paste(paste(Working_dir, "Input_Files/HT-SELEX_data_HOXD13", sep = "/"), "Hoxd13_TGATAA20NATA_AC_1.fastq.txt", sep = "/"), header = FALSE))
  }
  if(Select_cycle == "cycle2")
  {
    HOXD13_CHIP_tmp <- unique(read.table(paste(paste(Working_dir, "Input_Files/HT-SELEX_data_HOXD13", sep = "/"), "Hoxd13_TGATAA20NATA_AC_2.fastq.txt", sep = "/"), header = FALSE))
  }
  if(Select_cycle == "cycle3")
  {
    HOXD13_CHIP_tmp <- unique(read.table(paste(paste(Working_dir, "Input_Files/HT-SELEX_data_HOXD13", sep = "/"), "Hoxd13_TGATAA20NATA_AC_3.fastq.txt", sep = "/"), header = FALSE))
  }
  if(Select_cycle == "cycle4")
  {
    HOXD13_CHIP_tmp <- unique(read.table(paste(paste(Working_dir, "Input_Files/HT-SELEX_data_HOXD13", sep = "/"), "Hoxd13_TGATAA20NATA_AC_4.fastq.txt", sep = "/"), header = FALSE))
  }
  
  #
  HOXD13_CHIP <- as.data.frame(HOXD13_CHIP_tmp[sample(1:nrow(HOXD13_CHIP_tmp), 20000), ])
  colnames(HOXD13_CHIP) <- c("V1")
  
  # Obtain tiling
  numPeaks <- nrow(HOXD13_CHIP)
  Heatmap_DF_log2FC <- matrix(0, nrow = numPeaks, ncol = 13)
  Heatmap_DF_MPRAactivity <- matrix(0, nrow = numPeaks, ncol = 13)
  for(i in 1:numPeaks)
  {
   CurrentPeak <- toupper(HOXD13_CHIP$V1[i])
   Sequence_10mers_DF <- data.frame()
   for(j in 1:13)
   {
     Sequence_10mers_DF[j, 1] <- substr(CurrentPeak, j, j+7)
   }
   dds_results_HOXD13_df_subset <- dds_results_HOXD13_df[which(dds_results_HOXD13_df$TFBS %in% Sequence_10mers_DF$V1),]
   dds_results_HOXD13_df_subset <- dds_results_HOXD13_df_subset[match(Sequence_10mers_DF$V1, dds_results_HOXD13_df_subset$TFBS),]
   Heatmap_DF_log2FC[i,] <- dds_results_HOXD13_df_subset$log2FoldChange
   Heatmap_DF_MPRAactivity[i,] <- dds_results_HOXD13_df_subset$MPRAactivity
  }
  
  # Write
  if(Select_cycle == "input")
  {
    write.table(Heatmap_DF_log2FC, file = paste(paste(Working_dir, "Output_Files", sep = "/"), "Heatmap_DF_log2FC_8bpstiling_HOXD13_SELEX_input.txt", sep = "/"), col.names = FALSE, row.names = FALSE, quote = FALSE, sep = "\t")
    write.table(Heatmap_DF_MPRAactivity, file = paste(paste(Working_dir, "Output_Files", sep = "/"), "Heatmap_DF_MPRAactivity_8bpstiling_HOXD13_SELEX_input.txt", sep = "/"), col.names = FALSE, row.names = FALSE, quote = FALSE, sep = "\t")
  }
  if(Select_cycle == "cycle1")
  {
    write.table(Heatmap_DF_log2FC, file = paste(paste(Working_dir, "Output_Files", sep = "/"), "Heatmap_DF_log2FC_8bpstiling_HOXD13_HT-SELEX_cycle1.txt", sep = "/"), col.names = FALSE, row.names = FALSE, quote = FALSE, sep = "\t")
    write.table(Heatmap_DF_MPRAactivity, file = paste(paste(Working_dir, "Output_Files", sep = "/"), "Heatmap_DF_MPRAactivity_8bpstiling_HOXD13_HT-SELEX_cycle1.txt", sep = "/"), col.names = FALSE, row.names = FALSE, quote = FALSE, sep = "\t")
  }
  if(Select_cycle == "cycle2")
  {
    write.table(Heatmap_DF_log2FC, file = paste(paste(Working_dir, "Output_Files", sep = "/"), "Heatmap_DF_log2FC_8bpstiling_HOXD13_HT-SELEX_cycle2.txt", sep = "/"), col.names = FALSE, row.names = FALSE, quote = FALSE, sep = "\t")
    write.table(Heatmap_DF_MPRAactivity, file = paste(paste(Working_dir, "Output_Files", sep = "/"), "Heatmap_DF_MPRAactivity_8bpstiling_HOXD13_HT-SELEX_cycle2.txt", sep = "/"), col.names = FALSE, row.names = FALSE, quote = FALSE, sep = "\t")
  }
  if(Select_cycle == "cycle3")
  {
    write.table(Heatmap_DF_log2FC, file = paste(paste(Working_dir, "Output_Files", sep = "/"), "Heatmap_DF_log2FC_8bpstiling_HOXD13_HT-SELEX_cycle3.txt", sep = "/"), col.names = FALSE, row.names = FALSE, quote = FALSE, sep = "\t")
    write.table(Heatmap_DF_MPRAactivity, file = paste(paste(Working_dir, "Output_Files", sep = "/"), "Heatmap_DF_MPRAactivity_8bpstiling_HOXD13_HT-SELEX_cycle3.txt", sep = "/"), col.names = FALSE, row.names = FALSE, quote = FALSE, sep = "\t")
  }
  if(Select_cycle == "cycle4")
  {
    write.table(Heatmap_DF_log2FC, file = paste(paste(Working_dir, "Output_Files", sep = "/"), "Heatmap_DF_log2FC_8bpstiling_HOXD13_HT-SELEX_cycle4.txt", sep = "/"), col.names = FALSE, row.names = FALSE, quote = FALSE, sep = "\t")
    write.table(Heatmap_DF_MPRAactivity, file = paste(paste(Working_dir, "Output_Files", sep = "/"), "Heatmap_DF_MPRAactivity_8bpstiling_HOXD13_HT-SELEX_cycle4.txt", sep = "/"), col.names = FALSE, row.names = FALSE, quote = FALSE, sep = "\t")
  }
}

```


# Create stacked histogram (HT-SELEX)
```{r}
######################################################### Read files with tiled scores
# Input
Heatmap_DF_log2FC_input <- read.table(paste(paste(Working_dir, "Output_Files", sep = "/"), "Heatmap_DF_log2FC_8bpstiling_HOXD13_SELEX_input.txt", sep = "/"))
Heatmap_DF_MPRAactivity_input <- read.table(paste(paste(Working_dir, "Output_Files", sep = "/"), "Heatmap_DF_MPRAactivity_8bpstiling_HOXD13_SELEX_input.txt", sep = "/"))
Heatmap_DF_MPRAactivity_input[is.na(Heatmap_DF_MPRAactivity_input)] <- 0
Heatmap_DF_MPRAactivity_input[] <- lapply(Heatmap_DF_MPRAactivity_input, function(col) replace(col, col == -1, 0))

# Cycle-1
Heatmap_DF_log2FC_cycle1 <- read.table(paste(paste(Working_dir, "Output_Files", sep = "/"), "Heatmap_DF_log2FC_8bpstiling_HOXD13_HT-SELEX_cycle1.txt", sep = "/"))
Heatmap_DF_MPRAactivity_cycle1 <- read.table(paste(paste(Working_dir, "Output_Files", sep = "/"), "Heatmap_DF_MPRAactivity_8bpstiling_HOXD13_HT-SELEX_cycle1.txt", sep = "/"))
Heatmap_DF_MPRAactivity_cycle1[is.na(Heatmap_DF_MPRAactivity_cycle1)] <- 0
Heatmap_DF_MPRAactivity_cycle1[] <- lapply(Heatmap_DF_MPRAactivity_cycle1, function(col) replace(col, col == -1, 0))

# Cycle-2
Heatmap_DF_log2FC_cycle2 <- read.table(paste(paste(Working_dir, "Output_Files", sep = "/"), "Heatmap_DF_log2FC_8bpstiling_HOXD13_HT-SELEX_cycle2.txt", sep = "/"))
Heatmap_DF_MPRAactivity_cycle2 <- read.table(paste(paste(Working_dir, "Output_Files", sep = "/"), "Heatmap_DF_MPRAactivity_8bpstiling_HOXD13_HT-SELEX_cycle2.txt", sep = "/"))
Heatmap_DF_MPRAactivity_cycle2[is.na(Heatmap_DF_MPRAactivity_cycle2)] <- 0
Heatmap_DF_MPRAactivity_cycle2[] <- lapply(Heatmap_DF_MPRAactivity_cycle2, function(col) replace(col, col == -1, 0))

# Cycle-3
Heatmap_DF_log2FC_cycle3 <- read.table(paste(paste(Working_dir, "Output_Files", sep = "/"), "Heatmap_DF_log2FC_8bpstiling_HOXD13_HT-SELEX_cycle3.txt", sep = "/"))
Heatmap_DF_MPRAactivity_cycle3 <- read.table(paste(paste(Working_dir, "Output_Files", sep = "/"), "Heatmap_DF_MPRAactivity_8bpstiling_HOXD13_HT-SELEX_cycle3.txt", sep = "/"))
Heatmap_DF_MPRAactivity_cycle3[is.na(Heatmap_DF_MPRAactivity_cycle3)] <- 0
Heatmap_DF_MPRAactivity_cycle3[] <- lapply(Heatmap_DF_MPRAactivity_cycle3, function(col) replace(col, col == -1, 0))

# Cycle-4
Heatmap_DF_log2FC_cycle4 <- read.table(paste(paste(Working_dir, "Output_Files", sep = "/"), "Heatmap_DF_log2FC_8bpstiling_HOXD13_HT-SELEX_cycle4.txt", sep = "/"))
Heatmap_DF_MPRAactivity_cycle4 <- read.table(paste(paste(Working_dir, "Output_Files", sep = "/"), "Heatmap_DF_MPRAactivity_8bpstiling_HOXD13_HT-SELEX_cycle4.txt", sep = "/"))
Heatmap_DF_MPRAactivity_cycle4[is.na(Heatmap_DF_MPRAactivity_cycle4)] <- 0
Heatmap_DF_MPRAactivity_cycle4[] <- lapply(Heatmap_DF_MPRAactivity_cycle4, function(col) replace(col, col == -1, 0))

######################################################### Max consecutive 1s
Max_Lengths_input <- apply(Heatmap_DF_MPRAactivity_input, 1, function(col) Longest_1s(col)$max_length)
Max_Lengths_cycle1 <- apply(Heatmap_DF_MPRAactivity_cycle1, 1, function(col) Longest_1s(col)$max_length)
Max_Lengths_cycle2 <- apply(Heatmap_DF_MPRAactivity_cycle2, 1, function(col) Longest_1s(col)$max_length)
Max_Lengths_cycle3 <- apply(Heatmap_DF_MPRAactivity_cycle3, 1, function(col) Longest_1s(col)$max_length)
Max_Lengths_cycle4 <- apply(Heatmap_DF_MPRAactivity_cycle4, 1, function(col) Longest_1s(col)$max_length)

######################################################### Stacked Histogram
# 
GreaterThanWhat <- 6
Max_Lengths_input[which(Max_Lengths_input > GreaterThanWhat)] <- GreaterThanWhat
Max_Lengths_cycle1[which(Max_Lengths_cycle1 > GreaterThanWhat)] <- GreaterThanWhat
Max_Lengths_cycle2[which(Max_Lengths_cycle2 > GreaterThanWhat)] <- GreaterThanWhat
Max_Lengths_cycle3[which(Max_Lengths_cycle3 > GreaterThanWhat)] <- GreaterThanWhat
Max_Lengths_cycle4[which(Max_Lengths_cycle4 > GreaterThanWhat)] <- GreaterThanWhat

# 
num_reads_per_category <- c(table(Max_Lengths_input), 
                            table(Max_Lengths_cycle1), 
                            table(Max_Lengths_cycle2),
                            table(Max_Lengths_cycle3),
                            table(Max_Lengths_cycle4))

# Create DF
stacked_barplot_df <- as.data.frame(num_reads_per_category)
stacked_barplot_df$numOverlappingBS <- names(num_reads_per_category)
stacked_barplot_df$Cycles <- c(rep("Cycle_0", times = length(table(Max_Lengths_input))),
                               rep("Cycle_1", times = length(table(Max_Lengths_cycle1))),
                               rep("Cycle_2", times = length(table(Max_Lengths_cycle2))),
                               rep("Cycle_3", times = length(table(Max_Lengths_cycle3))),
                               rep("Cycle_4", times = length(table(Max_Lengths_cycle4))))

# Filter out reads with 0 binding sites?
stacked_barplot_df <- stacked_barplot_df[which(stacked_barplot_df$numOverlappingBS != 0),]
stacked_barplot_df$num_reads_per_category[which(stacked_barplot_df$Cycles == "Cycle_0")] <- stacked_barplot_df$num_reads_per_category[which(stacked_barplot_df$Cycles == "Cycle_0")]/sum(stacked_barplot_df$num_reads_per_category[which(stacked_barplot_df$Cycles == "Cycle_0")])
stacked_barplot_df$num_reads_per_category[which(stacked_barplot_df$Cycles == "Cycle_1")] <- stacked_barplot_df$num_reads_per_category[which(stacked_barplot_df$Cycles == "Cycle_1")]/sum(stacked_barplot_df$num_reads_per_category[which(stacked_barplot_df$Cycles == "Cycle_1")])
stacked_barplot_df$num_reads_per_category[which(stacked_barplot_df$Cycles == "Cycle_2")] <- stacked_barplot_df$num_reads_per_category[which(stacked_barplot_df$Cycles == "Cycle_2")]/sum(stacked_barplot_df$num_reads_per_category[which(stacked_barplot_df$Cycles == "Cycle_2")])
stacked_barplot_df$num_reads_per_category[which(stacked_barplot_df$Cycles == "Cycle_3")] <- stacked_barplot_df$num_reads_per_category[which(stacked_barplot_df$Cycles == "Cycle_3")]/sum(stacked_barplot_df$num_reads_per_category[which(stacked_barplot_df$Cycles == "Cycle_3")])
stacked_barplot_df$num_reads_per_category[which(stacked_barplot_df$Cycles == "Cycle_4")] <- stacked_barplot_df$num_reads_per_category[which(stacked_barplot_df$Cycles == "Cycle_4")]/sum(stacked_barplot_df$num_reads_per_category[which(stacked_barplot_df$Cycles == "Cycle_4")])

# Make plot
ggplot(stacked_barplot_df, aes(fill=numOverlappingBS, y=num_reads_per_category, x=Cycles)) + 
  geom_bar(position="stack", stat="identity") + 
  scale_fill_viridis(discrete = T) 

```



# Consecutive overlapping binding sites versus 60-mer PBM signal intensity
```{r}
# Read 60-mer PBM signal intensities
HOXD13_PBM_All60mers <- read.delim(paste(paste(Working_dir, "Input_Files/PBM_data_HOXD13", sep = "/"), "HOXD13_REF_R2_alldata.txt", sep = "/"), header = TRUE, sep = "\t")
HOXD13_PBM_All60mers <- HOXD13_PBM_All60mers[complete.cases(HOXD13_PBM_All60mers),]

######################################################### Tile probes
HOXD13_PBM_All60mers$Longest_1s <- 0
HOXD13_PBM_All60mers$SecondLongest_1s <- 0
HOXD13_PBM_All60mers$ZeroesBetween <- -1
for(i in 1:nrow(HOXD13_PBM_All60mers))
{
  Current_DNA_sequence <- HOXD13_PBM_All60mers$Sequence[i]
  Sequence_8mers_DF <- data.frame()
  for(j in 1:(nchar(Current_DNA_sequence) - 7))
  {
    Sequence_8mers_DF[j, 1] <- substr(Current_DNA_sequence, j, j+7)
  }
  dds_results_HOXD13_df_subset <- dds_results_HOXD13_df[which(dds_results_HOXD13_df$TFBS %in% Sequence_8mers_DF$V1),]
  dds_results_HOXD13_df_subset <- dds_results_HOXD13_df_subset[match(Sequence_8mers_DF$V1, dds_results_HOXD13_df_subset$TFBS),]
  HOXD13_PBM_All60mers$num_active[i] <- sum(dds_results_HOXD13_df_subset$MPRAactivity)
  HOXD13_PBM_All60mers$sum_log2FC[i] <- sum(2^dds_results_HOXD13_df_subset$log2FoldChange)
  HOXD13_PBM_All60mers$max_log2FC_active_only[i] <- max(2^dds_results_HOXD13_df_subset$log2FoldChange)
  if(sum(dds_results_HOXD13_df_subset$MPRAactivity) == 0)
  {
    HOXD13_PBM_All60mers$sum_log2FC_active_only[i] <- 0
    HOXD13_PBM_All60mers$max_log2FC_active_only[i] <- 0
    next
  }
  HOXD13_PBM_All60mers$sum_log2FC_active_only[i] <- sum(2^dds_results_HOXD13_df_subset$log2FoldChange[which(dds_results_HOXD13_df_subset$MPRAactivity == 1)])
  
  # Longest and second longest 1's
  rle_1s <- rle(dds_results_HOXD13_df_subset$MPRAactivity)
  longest_second_vec <- sort(rle_1s$lengths[which(rle_1s$values == 1)], decreasing = TRUE)
  if(length(longest_second_vec) > 0)
  {
    HOXD13_PBM_All60mers$Longest_1s[i] <- longest_second_vec[1]
  }
  if(length(longest_second_vec) > 1)
  {
    HOXD13_PBM_All60mers$SecondLongest_1s[i] <- longest_second_vec[2]
    
    # Distance between the highest and second highest
    longest_index <- which.max(rle_1s$lengths)
    second_index <- (sort(rle_1s$lengths, decreasing = TRUE))[2]
  
    if(longest_index < second_index) {
    # Second highest to the right
    zeros_between <- sum(rle_1s$lengths[(longest_index+1):(second_index-1)]) 
    } else {
    # Second highest to the left 
    zeros_between <- sum(rle_1s$lengths[(second_index+1):(longest_index-1)])
    }
    HOXD13_PBM_All60mers$ZeroesBetween[i] <- zeros_between
  }
}

######################################################### Fit linear models
# subset to make sure there aren't any other active 9-mers
HOXD13_PBM_All60mers_subset <- HOXD13_PBM_All60mers[which(HOXD13_PBM_All60mers$num_active == HOXD13_PBM_All60mers$Longest_1s),]
# HOXD13_PBM_All60mers_subset <- HOXD13_PBM_All60mers_subset[which(HOXD13_PBM_All60mers_subset$num_active > 0), ]

# run lm
summary(lm(log10(HOXD13_PBM_All60mers_subset$Alexa488Adjusted) ~ HOXD13_PBM_All60mers_subset$sum_log2FC_active_only + HOXD13_PBM_All60mers_subset$max_log2FC_active_only))
summary(lm(log10(HOXD13_PBM_All60mers_subset$Alexa488Adjusted) ~ HOXD13_PBM_All60mers_subset$num_active + HOXD13_PBM_All60mers_subset$max_log2FC_active_only))

# relative importance
library(relaimpo)
model <- lm(log10(Alexa488Adjusted) ~ max_log2FC_active_only + num_active, data = HOXD13_PBM_All60mers_subset)
model <- lm(log10(Alexa488Adjusted) ~ max_log2FC_active_only + sum_log2FC_active_only, data = HOXD13_PBM_All60mers_subset)
relative_importance <- calc.relimp(model, type = "lmg", rela = TRUE)
relative_importance

######################################################### Boxplot - 1
# subset to make sure there aren't any other active 8-mers
HOXD13_PBM_All60mers_subset <- HOXD13_PBM_All60mers[which(HOXD13_PBM_All60mers$num_active == HOXD13_PBM_All60mers$Longest_1s),]
HOXD13_PBM_All60mers_subset$num_active[which(HOXD13_PBM_All60mers_subset$num_active > 7)] <- 7
table(HOXD13_PBM_All60mers_subset$num_active)
table(HOXD13_PBM_All60mers$num_active)

# Boxplot
boxplot(log10(HOXD13_PBM_All60mers_subset$Alexa488Adjusted) ~ HOXD13_PBM_All60mers_subset$num_active, 
        notch = TRUE, outline = FALSE)
pairwise.wilcox.test(HOXD13_PBM_All60mers_subset$Alexa488Adjusted,
                     HOXD13_PBM_All60mers_subset$num_active)

######################################################### Boxplot - 2
# subset
HOXD13_PBM_All60mers_subset <- HOXD13_PBM_All60mers[which(HOXD13_PBM_All60mers$Longest_1s + HOXD13_PBM_All60mers$SecondLongest_1s == 6), ]
HOXD13_PBM_All60mers_subset$interaction_term <- interaction(HOXD13_PBM_All60mers_subset$Longest_1s, 
                                                            HOXD13_PBM_All60mers_subset$SecondLongest_1s, 
                                                            drop = TRUE, lex.order = TRUE)

# Boxplot
boxplot(log10(HOXD13_PBM_All60mers_subset$Alexa488Adjusted) ~ droplevels(HOXD13_PBM_All60mers_subset$interaction_term), 
        notch = TRUE, outline = FALSE)
table(HOXD13_PBM_All60mers_subset$interaction_term)
pairwise.wilcox.test(HOXD13_PBM_All60mers_subset$Alexa488Adjusted,
                     HOXD13_PBM_All60mers_subset$interaction_term)

######################################################### Individual
kmer <- "ATTAAACA"
kmer_RC <- "TGTTTAAT"

# subset & filter
HOXD13_PBM_All60mers_subset <- HOXD13_PBM_All60mers[grepl(kmer, HOXD13_PBM_All60mers$Sequence, fixed = TRUE) | grepl(kmer_RC, HOXD13_PBM_All60mers$Sequence, fixed = TRUE), ]
HOXD13_PBM_All60mers_subset <- HOXD13_PBM_All60mers_subset[which(HOXD13_PBM_All60mers_subset$Longest_1s == HOXD13_PBM_All60mers_subset$num_active), ]

# plot
boxplot(HOXD13_PBM_All60mers_subset$Alexa488Adjusted ~ HOXD13_PBM_All60mers_subset$Longest_1s)
table(HOXD13_PBM_All60mers_subset$Longest_1s)
pairwise.wilcox.test(HOXD13_PBM_All60mers_subset$Alexa488Adjusted, HOXD13_PBM_All60mers_subset$Longest_1s)

```

# Consecutive overlapping 8-mer E-scores versus 10-mer PADIT-seq log2foldchange
```{r}
# Read HOXD13 10-mer PADIT-seq data
dds_results_HOXD13_df <- read.table(paste(paste(Working_dir, "Output_Files", sep = "/"), "dds_results_HOXD13_df.txt", sep = "/"), header = TRUE)
dds_results_HOXD13_df$MPRAactivity <- (dds_results_HOXD13_df$log2FoldChange > 0 & dds_results_HOXD13_df$padj < Cutoff_Value)*1

# Read and format HOXD13 PBM data
HOXD13_PBM_Escores <- read.table(paste(paste(Working_dir, "Input_Files/PBM_data_HOXD13", sep = "/"), "HOXD13_REF_R2_8mers_11111111.txt", sep = "/"), header = TRUE)
TEMP1 <- HOXD13_PBM_Escores[,c("X8.mer", "E.score")]
TEMP2 <- HOXD13_PBM_Escores[,c("X8.mer.1", "E.score")]
colnames(TEMP1) <- c("X8.mer", "E.score")
colnames(TEMP2) <- c("X8.mer", "E.score")
TEMP <- rbind(TEMP1, TEMP2)
HOXD13_PBM_Escores <- TEMP

# 
dds_results_HOXD13_df$mer8_1 <- substr(rownames(dds_results_HOXD13_df), 1, 8)
dds_results_HOXD13_df$mer8_2 <- substr(rownames(dds_results_HOXD13_df), 2, 9)
dds_results_HOXD13_df$mer8_3 <- substr(rownames(dds_results_HOXD13_df), 3, 10)

# Merge
TEMP1 <- merge(dds_results_HOXD13_df,
               HOXD13_PBM_Escores, 
               by.x = "mer8_1", by.y = "X8.mer")
colnames(TEMP1)[ncol(TEMP1)] <- "E.score_1"
TEMP2 <- merge(TEMP1,
               HOXD13_PBM_Escores, 
               by.x = "mer8_2", by.y = "X8.mer")
colnames(TEMP2)[ncol(TEMP2)] <- "E.score_2"
TEMP3 <- merge(TEMP2,
               HOXD13_PBM_Escores, 
               by.x = "mer8_3", by.y = "X8.mer")
colnames(TEMP3)[ncol(TEMP3)] <- "E.score_3"
dds_results_HOXD13_df_tiledEscores <- TEMP3

################################################################# Effect of flanking nucleotides (aggregate analysis)
correlation_test_results_df <- data.frame()
Unique_8mers <- unique(c(HOXD13_PBM_Escores_PADIT$X8.mer_1_8[which(HOXD13_PBM_Escores_PADIT$MPRAactivity == 1)], 
                         HOXD13_PBM_Escores_PADIT$X8.mer.1_1_8[which(HOXD13_PBM_Escores_PADIT$MPRAactivity == 1)]))

for(i in 1:length(Unique_8mers))
{
  dds_results_HOXD13_df_tiledEscores_subset <- dds_results_HOXD13_df_tiledEscores[which(dds_results_HOXD13_df_tiledEscores$mer8_2 %in% Unique_8mers[i]),]
  if(nrow(dds_results_HOXD13_df_tiledEscores_subset) < 8)
  {
    correlation_test_results_df[i, 1] <- 0
    correlation_test_results_df[i, 2] <- 1
    correlation_test_results_df[i, 3] <- 0
    correlation_test_results_df[i, 4] <- 0
    next
  }
  correlation_test_results <- cor.test(apply(dds_results_HOXD13_df_tiledEscores_subset[, c("E.score_1", "E.score_3")], 1, mean), 
                                       dds_results_HOXD13_df_tiledEscores_subset$log2FoldChange, 
                                       method = "pearson") 
  correlation_test_results_df[i, 1] <- correlation_test_results$estimate
  correlation_test_results_df[i, 2] <- correlation_test_results$p.value
  correlation_test_results_df[i, 3] <- nrow(dds_results_HOXD13_df_tiledEscores_subset)
  correlation_test_results_df[i, 4] <- sum((dds_results_HOXD13_df_tiledEscores_subset$MPRAactivity == 1)*1)
}
rownames(correlation_test_results_df) <- Unique_8mers
colnames(correlation_test_results_df) <- c("correlation_coeff", "p.value", "numObs", "numSig")

# Add PBM Escores to correlation_test_results_df
HOXD13_PBM_Escores <- unique(HOXD13_PBM_Escores)
TMP <- merge(correlation_test_results_df, 
              HOXD13_PBM_Escores, 
              by.x = 0, 
              by.y = "X8.mer")
rownames(TMP) <- TMP$Row.names
correlation_test_results_df <- TMP[, -1]; rm(TMP)

# 
correlation_test_results_df <- correlation_test_results_df[which(correlation_test_results_df$numObs > 7),]
correlation_test_results_df <- correlation_test_results_df[which(correlation_test_results_df$numSig > 0),]
correlation_test_results_df$p.adj <- p.adjust(correlation_test_results_df$p.value, 
                                              method = "fdr")
correlation_test_results_df$p.adj.binary <- (correlation_test_results_df$p.adj<0.1)*1
table(correlation_test_results_df$p.adj.binary)

# Histogram - 1
hist(correlation_test_results_df$correlation_coeff, #[which(correlation_test_results_df$E.score > 0.4)], 
     breaks = 100, freq = FALSE,
     xlim = c(-1, 1), 
     col = "dark green",
     xlab = "Pearson R (mean E-scores of flanking nucleotides VS HOXD13 PADIT-seq log2FC)", 
     main = "Effect of flanking nucleotides on HOXD13 PADIT-seq log2FC")

################################################################# Effect of flanking nucleotides (case by case basis)
kmers_to_highlight <- c("ACTTTACT")
dds_results_HOXD13_df_tiledEscores_subset2 <- dds_results_HOXD13_df_tiledEscores[which(dds_results_HOXD13_df_tiledEscores$mer8_2 == "ACTTTACT"),] 
dds_results_HOXD13_df_tiledEscores_subset2$Colour <- "black"
dds_results_HOXD13_df_tiledEscores_subset2$Colour[which(dds_results_HOXD13_df_tiledEscores_subset2$MPRAactivity == 1)] <- "red"
plot(apply(dds_results_HOXD13_df_tiledEscores_subset2[, c("E.score_1", "E.score_3")], 1, mean), 
     dds_results_HOXD13_df_tiledEscores_subset2$log2FoldChange,
     cex = 1.5, pch = 19, 
     col = dds_results_HOXD13_df_tiledEscores_subset2$Colour, 
     # xlim = c(0, 0.5), ylim = c(0, 3), 
     xlab = "Mean E-score", ylab = "PADIT-seq log2FC")
cor.test(apply(dds_results_HOXD13_df_tiledEscores_subset2[, c("E.score_1", "E.score_3")], 1, mean), 
         dds_results_HOXD13_df_tiledEscores_subset2$log2FoldChange)

################################################################# Linear model
# Subset
Unique_8mers <- unique(c(HOXD13_PBM_Escores_PADIT$X8.mer_1_8[which(HOXD13_PBM_Escores_PADIT$MPRAactivity == 1)], 
                         HOXD13_PBM_Escores_PADIT$X8.mer.1_1_8[which(HOXD13_PBM_Escores_PADIT$MPRAactivity == 1)]))
dds_results_HOXD13_df_tiledEscores_subset3 <- dds_results_HOXD13_df_tiledEscores[which(dds_results_HOXD13_df_tiledEscores$mer8_1 %in% Unique_8mers | 
                                                                                       dds_results_HOXD13_df_tiledEscores$mer8_2 %in% Unique_8mers | 
                                                                                       dds_results_HOXD13_df_tiledEscores$mer8_3 %in% Unique_8mers),] 

# lm
summary(lm(dds_results_HOXD13_df_tiledEscores_subset3$log2FoldChange ~ 
             dds_results_HOXD13_df_tiledEscores_subset3$E.score_1 + 
             dds_results_HOXD13_df_tiledEscores_subset3$E.score_2 + 
             dds_results_HOXD13_df_tiledEscores_subset3$E.score_3))

summary(lm(dds_results_HOXD13_df_tiledEscores_subset3$log2FoldChange ~ 
             dds_results_HOXD13_df_tiledEscores_subset3$E.score_1))

summary(lm(dds_results_HOXD13_df_tiledEscores_subset3$log2FoldChange ~ 
             dds_results_HOXD13_df_tiledEscores_subset3$E.score_2))

summary(lm(dds_results_HOXD13_df_tiledEscores_subset3$log2FoldChange ~ 
             dds_results_HOXD13_df_tiledEscores_subset3$E.score_3))

```

